File: elixir/raindrops/lib/raindrops.ex
defmodule Raindrops do
  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    [3, 5, 7]
    |> Enum.filter(&(rem(number, &1) == 0))
    |> Enum.map(&sound/1)
    |> case do
      [] -> number
      sounds -> sounds
    end
    |> to_string()
  end

  defp sound(3), do: "Pling"
  defp sound(5), do: "Plang"
  defp sound(7), do: "Plong"
end

File: zig/raindrops/raindrops.zig
const std = @import("std");

pub fn convert(buffer: []u8, n: u32) []const u8 {
    var stream = std.io.fixedBufferStream(buffer);
    inline for (.{ 3, 5, 7 }) |factor| {
        const sound = switch (factor) {
            3 => "Pling",
            5 => "Plang",
            7 => "Plong",
            else => unreachable,
        };
        if (@rem(n, factor) == 0) {
            stream.writer().writeAll(sound) catch unreachable;
        }
    }
    if (stream.pos == 0) {
        stream.writer().print("{}", .{n}) catch unreachable;
    }
    return stream.getWritten();
}
